File: src/gradesta-manager/defaults.go
package main

import (
	pb "./pb"
	"reflect"
	"strconv"
	"strings"
)

var (
	serviceStateType = reflect.TypeOf(pb.ServiceState{})
	cellRuntimeType  = reflect.TypeOf(pb.CellRuntime{})
	cellType         = reflect.TypeOf(pb.Cell{})
	linkType         = reflect.TypeOf(pb.Link{})
)

func fid(field_name string, rtype reflect.Type) uint32 { // Get a protobuf field id
	field, _ := rtype.FieldByName(field_name)
	field_id, _ := strconv.Atoi(strings.Split(field.Tag.Get("protobuf"), ",")[1])
	return uint32(field_id)
}

func ss_fid(field_name string) uint32 { //ServiceState field id
	return fid(field_name, serviceStateType)
}

func cr_fid(field_name string) uint32 { //CellRuntime field id
	return fid(field_name, cellRuntimeType)
}

func c_fid(field_name string) uint32 { //Cell field id
	return fid(field_name, cellType)
}

func l_fid(field_name string) uint32 { //Link field id
	return fid(field_name, linkType)
}

var (
	default_state = pb.ClientState{
		ServiceState: &pb.ServiceState{
			OnDiskState: pb.ServiceState_READ_ONLY.Enum(),
			ServiceStateModes: map[uint32]*pb.Mode{
				ss_fid("Cells"):              &pb.Mode{Read: &truev, Write: &truev},
				ss_fid("NewCells"):           &pb.Mode{Read: &truev, Write: &truev},
				ss_fid("InView"):             &pb.Mode{Read: &truev, Write: &falsev},
				ss_fid("Index"):              &pb.Mode{Read: &truev, Write: &truev},
				ss_fid("OnDiskState"):        &pb.Mode{Read: &truev, Write: &truev},
				ss_fid("Round"):              &pb.Mode{Read: &truev, Write: &falsev},
				ss_fid("Log"):                &pb.Mode{Read: &truev, Write: &falsev},
				ss_fid("Metadata"):           &pb.Mode{Read: &truev, Write: &falsev},
				ss_fid("CellTemplate"):       &pb.Mode{Read: &truev, Write: &falsev},
				ss_fid("IdentityChallenge"):  &pb.Mode{Read: &truev, Write: &falsev},
				ss_fid("UserPublicKey"):      &pb.Mode{Read: &truev, Write: &falsev},
				ss_fid("UserSignature"):      &pb.Mode{Read: &truev, Write: &falsev},
				ss_fid("CapchaServers"):      &pb.Mode{Read: &truev, Write: &falsev},
				ss_fid("RequestedUserAttrs"): &pb.Mode{Read: &truev, Write: &falsev},
				ss_fid("UserAttrs"):          &pb.Mode{Read: &truev, Write: &falsev},
				ss_fid("ServiceStateModes"):  &pb.Mode{Read: &truev, Write: &falsev},
			},
			Round: &pb.Round{Request: &one},

			CellTemplate: &pb.CellRuntime{
				Cell: &pb.Cell{
					Encoding: pb.Encoding_UTF8.Enum(),
					Mime:     &text_plain,
				},
				UpdateCount: &one,
				ClickCount:  &one,
				Deleted:     &falsev,

				CellRuntimeModes: map[uint32]*pb.Mode{
					cr_fid("Cell"):               &pb.Mode{Read: &truev, Write: &truev},
					cr_fid("UpdateCount"):        &pb.Mode{Read: &truev, Write: &truev},
					cr_fid("ClickCount"):         &pb.Mode{Read: &truev, Write: &falsev},
					cr_fid("Deleted"):            &pb.Mode{Read: &truev, Write: &truev},
					cr_fid("CreationId"):         &pb.Mode{Read: &truev, Write: &falsev},
					cr_fid("CellRuntimeModes"):   &pb.Mode{Read: &truev, Write: &falsev},
					cr_fid("CellModes"):          &pb.Mode{Read: &truev, Write: &falsev},
					cr_fid("ForLinkModes"):       &pb.Mode{Read: &truev, Write: &falsev},
					cr_fid("BackLinkModes"):      &pb.Mode{Read: &truev, Write: &falsev},
					cr_fid("SupportedEncodings"): &pb.Mode{Read: &truev, Write: &falsev},
				},
				CellModes: map[uint32]*pb.Mode{
					c_fid("Data"):     &pb.Mode{Read: &truev, Write: &truev},
					c_fid("Encoding"): &pb.Mode{Read: &truev, Write: &falsev},
					c_fid("Mime"):     &pb.Mode{Read: &truev, Write: &falsev},
					c_fid("Forth"):    &pb.Mode{Read: &truev, Write: &truev},
					c_fid("Back"):     &pb.Mode{Read: &truev, Write: &truev},
					c_fid("Tags"):     &pb.Mode{Read: &truev, Write: &falsev},
					c_fid("Coords"):   &pb.Mode{Read: &truev, Write: &falsev},
				},
				LinkModes: map[uint32]*pb.Mode{
					l_fid("ServiceId"): &pb.Mode{Read: &truev, Write: &falsev},
					l_fid("Path"):      &pb.Mode{Read: &truev, Write: &falsev},
					l_fid("CellId"):    &pb.Mode{Read: &truev, Write: &truev},
				},
				ForLinkModes: map[uint64]*pb.Mode{
					0: &pb.Mode{Read: &truev, Write: &truev},
					1: &pb.Mode{Read: &truev, Write: &truev},
				},
				BackLinkModes: map[uint64]*pb.Mode{
					0: &pb.Mode{Read: &truev, Write: &truev},
					1: &pb.Mode{Read: &truev, Write: &truev},
				},
				SupportedEncodings: []pb.Encoding{pb.Encoding_UTF8},
			},
		},
	}
)

File: src/gradesta-manager/gradestalib/src/defaults.rs
use gradesta;
use std::collections::HashMap;

pub fn no_access_mode() -> gradesta::Mode {
  Default::default()
}

pub fn rw_mode() -> gradesta::Mode {
 gradesta::Mode {
  read: Some(true),
  write: Some(true),
  ..Default::default()
 }
}

pub fn ro_mode() -> gradesta::Mode {
 gradesta::Mode {
  read: Some(true),
  ..Default::default()
 }
}

pub fn default_cell_runtime_template() -> gradesta::CellRuntime {
 gradesta::CellRuntime {
  update_count: 0,
  click_count: Some(0),
  deleted: Some(false),
  cell_runtime_modes: hashmap! {
   1 => rw_mode(),
   2 => rw_mode(),
   3 => rw_mode(),
   4 => rw_mode(),
   5 => rw_mode(),
   6 => ro_mode(),
   7 => ro_mode(),
   8 => ro_mode(),
   9 => ro_mode(),
   10 => ro_mode()
  },
  cell_modes: hashmap! {
   1 => rw_mode(),
   2 => ro_mode(),
   3 => ro_mode(),
   4 => no_access_mode(),
   5 => rw_mode(),
   6 => rw_mode(),
   200 => no_access_mode()
  },
  link_modes: hashmap! {
   1 => no_access_mode(),
   2 => no_access_mode(),
   3 => rw_mode(),
   4 => no_access_mode()
  },
  link_direction_modes: hashmap! {
   1 => rw_mode(),
   -1 => rw_mode(),
   2 => rw_mode(),
   -2 => rw_mode()
  },
  ..Default::default()
 }
}

pub fn default_service_state() -> gradesta::ServiceState{
 gradesta::ServiceState {
  on_disk_state: Some(gradesta::service_state::OnDiskState::Saved as i32),
  cell_template: Some(default_cell_runtime_template()),
  service_state_modes: hashmap! {
   1 => rw_mode(),
   2 => rw_mode(),
   4 => ro_mode(),
   5 => ro_mode(),
   7 => ro_mode(),
   8 => ro_mode(),
   9 => ro_mode(),
   10 => no_access_mode(),
   11 => no_access_mode(),
   12 => no_access_mode(),
   13 => ro_mode()
  },
  ..Default::default()
 }
}
